lib/msf/core/db_manager/http/servlet/host_servlet.rb

```diff
@@ -28,13 +28,14 @@ def self.get_host
     lambda {
       warden.authenticate!
       begin
-        opts = parse_json_request(request, false)
         sanitized_params = sanitize_params(params)
         data = get_db.hosts(sanitized_params)
         includes = [:loots]
-        set_json_response(data, includes)
+        # Only return the single object if the user used the resource/ID GET request
+        data = data.first if data.count == 1 && request.url =~ /\/\d$/
```

Comment:
This change doesn't behave as I believe you intend. If we consider the example request `http://localhost:8080/api/v1/hosts/12?workspace=default`, `data.count == 1` evaluates to true, however `request.url =~ /\/\d$/` evaluates to nil, thus `data.count == 1 && request.url =~ /\/\d$/` results in a nil and the first element is not returned. Rewrite this and all other such instances as `data = data.first if !sanitized_params[:id].nil? && data.count == 1`.

Reply:
Yeah, the regex is broken. The reason I was going with a regex is because just checking if `:id` is present will cause `/resources?id=2` to return a single object instead of an array.  It is confusing as a consumer to have `/resources?query_params` sometimes return an array and sometimes return a single object. The consumer shouldn't have to verify the return type is one thing or the other in order to process the data.

For the sake of time, I will make your proposed suggestion. But I do think we should revisit in the future.

Reply:
I looked into it further and If we leverage an aspect of the Rack environment we can determine if the user provided the ID via URL vs the query string.
```
-        data = data.first if !sanitized_params[:id].nil? && data.count == 1
+        data = data.first if !sanitized_params[:id].nil? && !env['rack.request.query_hash'].key?('id') && data.count == 1
```

We can determine this via the `params` hash. When the user provides an ID via the URL the hash contains a `captures` key where the value is an Array of the captures. The lookup wouldn't be as simple if we consider the possibility of routes with additional captures in the future.

Example request `/api/v1/hosts/12`:
```
params = Hash (3 elements)
 'splat' = Empty Array
 'captures' = Array (1 element)
  [0] = "12"
 'id' => 12
```

Reply:
Fixed with the latest push. I also refactored the logic for determining if it should be a single object or not into a helper because changing it across 10ish servlets each time was getting annoying.